docs/06-contributors/999-rfcs/000-stories/story-13.md

```diff
@@ -0,0 +1,169 @@
+# User Story 13 - Task List with uuid, Json and duration
+
+> **Status**: Expected released on 2023/02/16
+
+The following code is an iteration of Wing's task list.
+This version of it includes the following functional changes:
+- Using npm's uuid package inflight instead of the counter resource 
+- Storing Json instead of a string
+- Using duration for estimated effort
+
+Also, it includes some non-functional requirements: 
+- The Bucket should be replicated across different regions 
+  - +100 if we are able to do the same for azure as well
+- Least privileged permissions are granted
+- The code should work on sim, aws, azure
+- VScode should be able to autocomplete 
+- Console should be able to show all resources, and interact with them
+
+
+
+```ts (wing)
+bring cloud;
+// need to run: npm install uuid --save
+bring uuid;
+
+/**
+ * Represents a cloud task list.
+ */
+resource TaskList {
+  /** stores the tasks */
+  _bucket: cloud.Bucket;
+  
+  init() {
+    this._bucket = new cloud.Bucket();
+  }
+
+  /** 
+   * Adds a task to the task list.
+   * @returns The ID of the new task.
+   */
+  inflight add_task(title: str): str {
+    let id = uuid.v4();
+    let j = Json { 
+      title: title, 
+    };
+    print("adding task ${id} with data: ${j}"); //j should be printed out nicely 
+    this._bucket.put(id, j.to_str());
+    return id;
+  }
+
+  /** 
+   * Gets a task from the task list.
+   * @param id - the id of the task to return
+   * @returns the title of the task (optimistic)
+   */
+  inflight get_task(id: str): Json {
+    let blob = this._bucket.get(id);
+    return Json.parse(blob)
+  }
+
+  /** 
+   * Sets effort estimation on a test
+   * @param id - the id of the task to return
+   * @param effort_estimation - the time (duration) estimated for this task
+   * @returns The ID of the existing task.
+   */
+  inflight add_estimation(id: str, effort_estimation: duration) {
+    let j = this.get_task(id).to_mut();
+    j.set("effort_estimation", effort_estimation);
+    this._bucket.put(id, j.to_str());
+    return id;
+  }
+
+  /** 
+   * Removes a task from the list
+   * @param id - the id of the task to be removed
+   */
+  inflight remove_tasks(id: str): str {
+    print("removing task ${id}");
+    this._bucket.delete(id);
+    return id;
+  }
+
+   /** 
+    * Gets the tasks ids 
+    * @returns set of task id
+    */
+  inflight list_task_ids(): Set<str> {
+    let result = MutSet<str> {};
+    for id in this._bucket.list() {
+      result.add(id);
+    }
+    return result.to_immut();
```

Comment:
```suggestion
    return result.clone();
```

Reply:
I find it counter intuitive that clone returns a different object type. 
I would expect reserving the `clone` for 🤷  cloning

Reply:
BTW, I also don't like the `to_immut()` API 😊

Reply:
I'm a fan of clone or copy. Maybe copy is better because there's less implication of "deep" copy (which is not what we're doing). It's also 1 fewer letter :)

`result.copy()` and `result.copy_mut()` are nice because the effect is clear, the fact that the type changes doesn't feel as important to put in the name because it's already in the signature.

Reply:
We are having two separate discussions here:

1. Should we call this method `copy` or `clone`?
2. What should be the name of the method that returns an immutable copy and a mutable one?

The rationale of having `copy`/`clone` return an immutable version is because we want Wing to be ["immutable by default"](https://docs.winglang.io/blog/2023/02/02/good-cognitive-friction). So I am actually comfortable with the fact that this is counter intuitive (this is what I was calling "good cognitive friction").

I am less comfortable with `clone()` returning a *mutable* copy because this means that users can mistakingly use a mutable object and we want that to be a very explicit thing.

So in my view, the alternative to `copy()`+`copy_mut()` are `copy_immut()`+`copy_mut()` (that is, no `clone()`), which TBH, is quite ugly.

Bottom line, I still think `clone()` (or `copy()`) should return the immutable version, and if the user is "surprised" and has to explicitly request the mutable copy, I am happy about it.

(I am less opinionated about `clone` versus `copy` :-))

Reply:
```suggestion
    return result.copy_mut();
```